From 22c5deddd3d57dedddd9c9dfc3aca395a2e32cf2 Mon Sep 17 00:00:00 2001
From: Wen He <wen.he_1@nxp.com>
Date: Thu, 29 Nov 2018 05:35:14 +0800
Subject: [PATCH 377/706] drm: hdp: add support swap lane mapping

Signed-off-by: Wen He <wen.he_1@nxp.com>
(cherry picked from commit 1d15ab55c829e28a7ff940ecc9e7ea6ea4daf582)
Signed-off-by: Zumeng Chen <zumeng.chen@windriver.com>
---
 drivers/gpu/drm/imx/hdp/imx-dp.c | 19 +++++++++++++++++++
 1 file changed, 19 insertions(+)

diff --git a/drivers/gpu/drm/imx/hdp/imx-dp.c b/drivers/gpu/drm/imx/hdp/imx-dp.c
index 1855bb6bfe79..0ae1e977496b 100644
--- a/drivers/gpu/drm/imx/hdp/imx-dp.c
+++ b/drivers/gpu/drm/imx/hdp/imx-dp.c
@@ -70,6 +70,16 @@ int dp_fw_init(state_struct *state)
 		DRM_ERROR("echo test failed, check firmware!\n");
 		return -ENXIO;
 	}
+	DRM_INFO("CDN_API_General_Test_Echo_Ext_blocking (ret = %d echo_resp = %s)\n",
+		ret, echo_resp);
+
+	/* Line swaping */
+	CDN_API_General_Write_Register_blocking(state,
+						ADDR_SOURCD_PHY +
+						(LANES_CONFIG << 2),
+						0x00400000 |
+						hdp->lane_mapping);
+	DRM_INFO("CDN_API_General_Write_Register_blockin ... setting LANES_CONFIG\n");
 
 	return 0;
 }
@@ -469,6 +479,7 @@ int dp_phy_init_t28hpc(state_struct *state,
 {
 	struct imx_hdp *hdp = state_to_imx_hdp(state);
 	int max_link_rate = hdp->link_rate;
+	int lane_mapping = hdp->lane_mapping;
 	int num_lanes = 4;
 	int ret;
 
@@ -483,6 +494,14 @@ int dp_phy_init_t28hpc(state_struct *state,
 		num_lanes = hdp->edp_num_lanes;
 	}
 
+	/* Line swaping */
+	CDN_API_General_Write_Register_blocking(state,
+						ADDR_SOURCD_PHY +
+						(LANES_CONFIG << 2),
+						0x00400000 | lane_mapping);
+	DRM_INFO("CDN_*_Write_Register_blocking ... setting LANES_CONFIG %x\n",
+		 lane_mapping);
+
 	/* PHY initialization while phy reset pin is active */
 #ifdef CONFIG_EMU_PXP
 	AFE_init(state, num_lanes, (ENUM_AFE_LINK_RATE)max_link_rate);
-- 
2.17.1

